### PR TITLE
Certificate management fixes

### PR DESCRIPTION
**Pull request by smarnach**
This PR contains a few fixes and improvements:

1. The haproxy-config script converts all domain names to lower case when applying a new backend map.  To be defensive, we also convert all domain names to lower case in the certificate management script; it shouldn't be possible for uppper-case domain names to get there anymore.  The integration test playbook now also tests with the domain name converted to upper case.

2. The certificate clean-up job so far only renamed the renewal configuration to disable certificate renewal.  This causes problems if the same domain gets re-enabled.  Since we have a backup of `/etc/letsencrypt`, it's ok to more thoroughly delete all files related to an unused certificate.  I've included a test for this in the integration test playbook. this

3. The `-expand` flag to letsencrypt doesn't seem to be working as advertised.  Whe requesting a certificate for the domains A and B, and then rerequesting for A, B and C, the `--expand` flag is supposed to issue a new certificate for the expanded set of domains.  However, letsencrypt instead simply assumes that the certificate isn't ready for renewal yet, and silently exits with a success exit code.  Adding the `--force-renew` flag works around this problem.  Since we only invoke letsencrypt when we have detected that a cert is missing, this should not result in redundant certificate requests.

4. Improved error logging for both `manage_certs.py` and `haproxy-reload`.  Both scripts now print errors and useful debugging info to stderr to make sure it gets send to the ops email address by cron.

5. Add some unit tests.

To test this, run the integration and unit tests.

### CODE REVIEW

File: templates/manage_certs.py
Comment (antoviaque) [inline templates/manage_certs.py:36]: Missing a test?

**Review by antoviaque — CHANGES_REQUESTED** [review]

File: templates/manage_certs.py
Comment (smarnach) [reply templates/manage_certs.py:36]: @antoviaque This is just defensive coding.  The actual bug was on the side of the instance manager, which needs to convert domain names to lower case.  Domain names with uppercase letters won't work correctly in the haproxy configuration, so testing this in the integration test playbook is very hard.  I will think more about this and see whether I can come up with an actually useful way of testing this.

Comment (smarnach) [reply templates/manage_certs.py:36]: @antoviaque I've modified the configuration management script to automatically convert all domain names to lower case for all new backend maps.  This is a better solution, since clients of the load balancer are no longer required to send lower case domain names.  This also makes it easy to test this in the integration test playbook.

File: tasks/main.yml
Comment (pomegranited) [inline tasks/main.yml:84]: I'm unclear on how the cron emails are configured to reach `ops@opencraft.com`.

Could you maybe set the MAILTO for the crontab to `LOAD_BALANCER_OPS_EMAIL`?

[Ansible 2.1 supports adding cron env variables](http://docs.ansible.com/ansible/cron_module.html#).

**Review by pomegranited — COMMENTED** [review]
@smarnach I'm waiting for DNS to propagate so I can run the integration playbook.

Could you provide instructions for running the playbook on e.g. haproxy-dev.opencraft.hosting, so I can test the effects on my devstack appserver activation?

I also have a question, so thought I'd forward it along now.

File: tasks/main.yml
Comment (smarnach) [reply tasks/main.yml:84]: Cron mails are sent to root by default, and when this role is included in our full setup, root email will be [forwarded to `OPS_EMAIL`](https://github.com/open-craft/ansible-forward-server-mail/blob/master/templates/aliases#L1) and [`LOAD_BALANCER_OPS_EMAIL` is also set to `OPS_EMAIL`](https://github.com/open-craft/deployment-deploy-services/blob/master/group_vars/load-balancer/public.yml#L11).  In the integration tests, the load-balancer role is used in isolation, so mail isn't forwarded.

**Comment by smarnach** [comment]
@pomegranited Instructions to deploy to a production load-balancing server are here: https://github.com/open-craft/deployment-deploy-services#perform-the-deployment

To limit the deployment to just `haproxy-dev`, use `-l haproxy-dev.opencraft.hosting` instead of `-l load-balancer` when running the playbook.

File: templates/manage_certs.py
Comment (antoviaque) [reply templates/manage_certs.py:36]: @smarnach I only meant to add a unit test in that case - although an integration test is definitely a plus.

Oh, wait - where are the unit tests for this python code? I only see integration tests for the playbook as a whole in https://github.com/open-craft/ansible-load-balancer/tree/master/tests

Comment (smarnach) [reply templates/manage_certs.py:36]: @antoviaque Most of the code can only be meaningfully unit tested by implementing a fake for the letsencrypt client, for haproxy, for the haproxy-config-watcher and the file system, due to the high level of integration of this code.  For this reason, I implemented integration tests with good test coverage instead, which also test all the other components of this role.  I think that this provides much more value for far less work.

I now added unit tests for a few functions, inlcuding this one, where we might get some limited value from such tests.

File: README.md
Comment (pomegranited) [inline README.md]: The integration tests will run within a python2 virtualenv, and error out with some ansible issue about "basestring" not being defined.

But to get the unit tests to work, I had to use a python3 virtualenv, and run `pip install pyopenssl`.

Using the python2 virtualenv, even after installing `pip install pyopenssl`, I kept getting this `ImportError` for OpenSSL, and couldn't make it go away.. does that seem right to you?

```
======================================================================
ERROR: test_manage_certs (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_manage_certs
Traceback (most recent call last):
  File "/usr/lib/python3.5/unittest/loader.py", line 153, in loadTestsFromName
    module = __import__(module_name)
  File "/home/jill/instance-manager/ansible-load-balancer/tests/test_manage_certs.py", line 15, in <module>
    import manage_certs
  File "/home/jill/instance-manager/ansible-load-balancer/templates/manage_certs.py", line 23, in <module>
    import OpenSSL.crypto
ImportError: No module named 'OpenSSL'
```

Please update this README and add `pyopenssl` to the `tests/requirements.txt`.

File: tasks/main.yml
Comment (pomegranited) [reply tasks/main.yml:84]: Thanks for the clarification, @smarnach .

File: tests/test_configuration.yml
Comment (pomegranited) [inline tests/test_configuration.yml]: My integration tests are timing out at this step.  It looks like the .pem file gets created fine, but it doesn't have the expected file name?

```
root@jill-test-haproxy:/etc/haproxy/certs# ll 
total 20
drwx------ 2 root root 4096 Nov 17 05:38 ./
drwxr-xr-x 6 root root 4096 Nov 17 05:05 ../
-rw-r--r-- 1 root root 5176 Nov 17 05:38 jill-test-haproxy.plebia.net-0001.pem
-rw-r--r-- 1 root root 2729 Nov 17 05:52 ssl-cert-snakeoil.pem
```

Note that:
- The first time I ran these tests on my brand new instance, I did it from `origin/master`.  Tests succeeded.
- The second time I ran them, I did it from your PR branch.  This failed with: 
  
  ```
  TASK [wait for the certificate to be available] ********************************
  fatal: [jill-test-haproxy.plebia.net]: FAILED! => {"changed": false, "elapsed": 180, "failed": true, "invocation": {"module_args": {"connect_timeout": 5, "delay": 0, "exclude_hosts": null, "host": "127.0.0.1", "path": "/etc/haproxy/certs/jill-test-haproxy.plebia.net.pem", "port": null, "search_regex": null, "state": "started", "timeout": 180}, "module_name": "wait_for"}, "msg": "Timeout when waiting for file /etc/haproxy/certs/jill-test-haproxy.plebia.net.pem"}
  ```
- The third time I ran them,  while ansible was waiting for the file to appear, I copied `jill-test-haproxy.plebia.net-0001.pem` to `jill-test-haproxy.plebia.net.pem` to get that task to succeed. But the tests went on to fail here:
  
  ```
  TASK [verify that the backend is accessible via HTTPS] *************************
  fatal: [jill-test-haproxy.plebia.net -> localhost]: FAILED! => {"changed": true, "cmd": ["wget", "-O-", "--ca-certificate=../files/lets-encrypt-fake-cert.crt", "https://jill-test-haproxy.plebia.net/"], "delta": "0:00:05.140661", "end": "2016-11-17 16:36:07.617502", "failed": true, "invocation": {"module_args": {"_raw_params": "wget -O- --ca-certificate=../files/lets-encrypt-fake-cert.crt \"https://jill-test-haproxy.plebia.net/\"", "_uses_shell": false, "chdir": null, "creates": null, "executable": null, "removes": null, "warn": true}, "module_name": "command"}, "rc": 8, "start": "2016-11-17 16:36:02.476841", "stderr": "--2016-11-17 16:36:02--  https://jill-test-haproxy.plebia.net/\nResolving jill-test-haproxy.plebia.net (jill-test-haproxy.plebia.net)... 213.32.78.69\nConnecting to jill-test-haproxy.plebia.net (jill-test-haproxy.plebia.net)|213.32.78.69|:443... connected.\nHTTP request sent, awaiting response... 503 Service Unavailable\n2016-11-17 16:36:07 ERROR 503: Service Unavailable.", "stdout": "", "stdout_lines": [], "warnings": ["Consider using get_url module rather than running wget"]}
  ```
- I reverted to `origin/master` and re-ran the tests, but got the same error at `wait for the certificate to be available`.

Is there some cleanup that needs to be done?  Or do I really have to run with a fresh instance each time?

**Review by pomegranited — CHANGES_REQUESTED** [review]
Sorry @smarnach, I encountered some issues when running the unit and integration tests.
- [-] I tested this by running the integration and unit tests on my dev load balancer instance.  See errors described inline.
- [x] I read through the code, and it looks fine.
- [x] <strike>I checked for accessibility issues<strike>
- [x] Includes documentation

File: tests/test_configuration.yml
Comment (smarnach) [reply tests/test_configuration.yml]: @pomegranited The version from `master` doesn't clean up old certificates properly, and the files that get left behind interfere with the next integration test run and cause the additional `-0001` in the file name.  (Note that this wouldn't be a problem in production if it happened; it's only a problem since we are waiting for a file with the exact name `<domain>.pem` in the integration test.  I would like to have a more flexible condition in the integration test, but Ansible's `wait_for` is rather limited.)

This problem has been fixed on this branch, but since you ran the playbook on master first, you still experienced it for now.  To make the tests more robust, I added some clean-up code to the playbook that removes any left-over files first.  This shouldn't be necessary in the future, but it doesn't hurt.

The second error indicates that the backend server, i.e. `pydoc`, has died in the meantime.  I've got no idea why this happened.  It is configured to run for ten minutes, which should be enough.  I simply upped this time to one hour;  pydoc is killed when Ansible closes the SSH connection anyway.

File: README.md
Comment (smarnach) [reply README.md]: @pomegranited For some reason I can run the unit tests from inside the Python 2 virtualenv just fine, so I didn't notice this.

I will clean this up a bit.  Adding pyopenssl to tests/requirements.txt doesn't work, since we need the Python 3 version of the module.

Comment (smarnach) [reply README.md]: @pomegranited I've added a Makefile and updated the instructions in the README.

**Review by pomegranited — APPROVED** [review]
👍 

Thank you @smarnach, that worked!  The tests run fine now.

And thank you for adding the `tests/Makefile`, that makes them really easy to use.

**Comment by smarnach** [comment]
@antoviaque I'm still waiting for your approval on this PR.

File: tests/Makefile
Comment (antoviaque) [inline tests/Makefile]: @smarnach We use prospector for OpenCraft IM - this would allow to have the same conventions/configuration for the linters on both sides.

Same thing for the `Makefile` rules - given that this is part of the same project, could you use the same rule names when appropriate?

File: tests/Makefile
Comment (antoviaque) [inline tests/Makefile]: @smarnach Could you also generate a coverage report? No need to make it fail below a given percentage for now, but it would help to see what is being run by the unit tests.

File: tests/Makefile
Comment (antoviaque) [inline tests/Makefile:41]: @smarnach And `make test` to run all tests & linters, and get circleci to run it automatically? That would save time for reviewers, and ensure we don't forget to run some of these at times.

File: templates/manage_certs.py
Comment (antoviaque) [reply templates/manage_certs.py:36]: @smarnach Thank you for the unit tests. That's true that for some of the functions, relying on the integration tests is enough, and will provide better guarantees than a unit test. Having both though helps to add granularity to the blanket testing from the integration tests, and having the infrastructure in place will helps others working on this in the future - when there isn't any, it becomes very tempting to not write unit tests.

File: tests/test_manage_certs.py
Comment (antoviaque) [inline tests/test_manage_certs.py]: @smarnach This is a case where writing an isolated test for this would really help to point out the issue, and have clearer errors if that test ever fails (since the name of the test would describe the bug). Since you've already written the tests all at once now, I won't ask you to change it here - but a comment to highlight the issue and not lose the context/test in the future would still be good. Could you consider it when fixing bugs in the future though?

**Review by antoviaque — CHANGES_REQUESTED** [review]

File: templates/manage_certs.py
Comment (smarnach) [reply templates/manage_certs.py:36]: @antoviaque Yup, all agreed, in particular about the importance of having the infrastructure in place.

File: tests/Makefile
Comment (smarnach) [reply tests/Makefile]: I've changed this to use prospector and renamed the rules.

Do you want me to copy the full pylint and prospector config from the instance manager?  It will likely get out of sync soon.

Comment (smarnach) [reply tests/Makefile]: As a side note in reply to "part of the same project", there is nothing in the load balancer role that is specific to the instance manager.  It can be useful wherever you want an LB that manages SSL certificates for you.

File: tests/Makefile
Comment (smarnach) [reply tests/Makefile:41]: @antoviaque The integration tests require a server and a DNS name pointing to that server.  If I want to provision this automatically on CircleCI, I would have to add infrastructure to create a VM and to access Gandi to set a DNS name, but even if I did this, the DNS name would not propagate before the tests time out.  So the options are

1. Use some alternative DNS infrastructure that allows for faster DNS propagation.  I'm privately running a bind server that I'm using for dynamic DNS, and updates are usually live and globally visible instantaneously for names that haven't been used recently.
2. Keep a VM constantly running for the integration tests.

The second option is a lot easier to implement, but OpenCraft would have to pay for the constantly running VM.  We can use the cheapest and smallest virtual server we can find.  It doesn't need to be on OpenStack.

File: tests/Makefile
Comment (antoviaque) [reply tests/Makefile]: @smarnach Thanks - and yes, it would be good to copy it from there. To avoid getting out of sync, maybe it could be done by wget'ing it from the master of the opencraft repo, this way it would not require duplicating the file in this repo?

Comment (antoviaque) [reply tests/Makefile]: And for the same project - yup, it's definitely a good feature that it could be used independently. It can still be considered as part of the project, though, it's not exclusive : )

File: tests/Makefile
Comment (antoviaque) [reply tests/Makefile:41]: @smarnach Thank you for figuring out the approach and options - the second one seem a lot faster and less brittle, so it's fine to use a continuously running server for this. Hopefully one day OVH will support floating IPs within OpenStack, and we can have better solutions for cases like this.

**Comment by antoviaque** [comment]
@smarnach Once the last points we are discussing are addressed, I'm all good on this. I won't need to review again. (Same thing as on the other PRs though, this isn't a :+1: as I haven't reviewed in details or tested - but since I believe you have only changed tests since Jill tested, a green build from CircleCI should be enough I think - if I'm wrong and this ends up being blocking for merging, ping me on IRC to make sure I see it quickly, I'll try to make time to help you merge this on time.)

**Review by antoviaque — APPROVED** [review]

File: tests/Makefile
Comment (smarnach) [reply tests/Makefile]: @antoviaque Several problems here.

1. If we use `wget` to download the file from the instance manager each time the tests are run then we can break the tests in this repo by making a change to the other repo.  I don't think we want this, so instead I suggest to add a make target that updates the files from the instance manager.  We will have to do this manually, but nothing bad happens if we don't do it, and it prevents our tests from randomly breaking.

2. The `pylintrc` from the instance manager can't be used unmodified, since it includes `load-plugins=lint_django`, which won't work for this repo.  I tried overriding this from the command line, but the plugins given on the command line are loaded in addition to those in the configuration file, so there is nothgin I can do.  This is another reason why we can't download the file from the instance manager.

3. I can't make prospector work when using the same configuration as the instance manager.  I runs fine and claims that all is good, but if I modify the files to contain pylint violations, they don't show up in the output.  If I run pylint manually, all works well, and when I add violations of e.g. pyflakes I can see them as well.  Prospector doesn't offer any debug output, so it's very hard to tell what's going on.  It _claims_ to use the correct pylint configuration file, but I can't see any indication that pylint is actually being run.  The next step here would be to dig into the prospector code to figure out what it's actually doing.

How much additional time am I supposed to invest in this?  I personally don't think the change to use exactly the same configuration adds much value over the current version, which is already mostly consistent with the instance manager, and we only have like 200 lines of code in this repo.

Comment (antoviaque) [reply tests/Makefile]: @smarnach 

> If we use wget to download the file from the instance manager each time the tests are run then we can break the tests in this repo by making a change to the other repo. I don't think we want this, so instead I suggest to add a make target that updates the files from the instance manager. We will have to do this manually, but nothing bad happens if we don't do it, and it prevents our tests from randomly breaking.

Yup, adding a make target seem fine yep - that would actually be in line with how we update dependencies. 

> The pylintrc from the instance manager can't be used unmodified, since it includes load-plugins=lint_django, which won't work for this repo. I tried overriding this from the command line, but the plugins given on the command line are loaded in addition to those in the configuration file, so there is nothgin I can do. This is another reason why we can't download the file from the instance manager.

With the make target, that could be something to `sed` out of the file you retrieve?

> I can't make prospector work when using the same configuration as the instance manager. I runs fine and claims that all is good, but if I modify the files to contain pylint violations, they don't show up in the output. If I run pylint manually, all works well, and when I add violations of e.g. pyflakes I can see them as well. Prospector doesn't offer any debug output, so it's very hard to tell what's going on. It claims to use the correct pylint configuration file, but I can't see any indication that pylint is actually being run. The next step here would be to dig into the prospector code to figure out what it's actually doing.
> How much additional time am I supposed to invest in this? I personally don't think the change to use exactly the same configuration adds much value over the current version, which is already mostly consistent with the instance manager, and we only have like 200 lines of code in this repo.

Since the LB is pretty much the first service outside of the instance manager repo, you're pretty much hitting all the bottlenecks; we should actually have thought of this during the discovery, I don't think we really planned for this. If that's some comfort, that will make it much easier for the next services to be setup, once you'll have done it once for here. 

That also makes it important to sort out now - it's precisely because it's only 200 lines of code for now that it's useful to do it now. If it grows, or when other services will be setup, it will be a higher barrier. Especially with things like linting, where adding it or changing it late can be especially painful/lengthy.

I would say it's worth digging a bit to get it to run. I don't want to make this spillover one more time though, so if you can't figure out why prospector doesn't run in the time you have available, I would be ok with creating a task for it on the scaling epic, but if we can avoid adding another tech debt story to the backlog that would be good.

Comment (smarnach) [reply tests/Makefile]: @antoviaque I finally got it working, but had to overcome a mental block.  I tried to answer the question why prospector does not work here, while it works fine for the instance manager.  The answer turned out to be that it currently isn't working for the instance manager as well, since the current version of prospector is incompatible with pylint 1.6.  Due to the opacity of prospector, we haven't noticed this when upgrading pylint to version 1.6.  I reverted the pylint version to 1.5.6 here, and now everything is working fine.  We need to fix this in the instance manager as well, but it is likely that we already have added a few pylint violations since it stopped working.  I'll take a look and create a ticket to fix this if appropriate.  (It could also be useful to get rid of prospector altogether, but I'll elaborate on the new ticket.)

Regarding the django plugin, another solution would be to remove it from the pylint configuration file and load it via a command-line flag to prospector in the instance manager.  This way, the confiugration files in both repos could be identical, and only the command-line invocations slightly different.  This seems a cleaner solution to me then trying to fix the configuration file with sed.

We definitely didn't plan for all this during the discovery, since at that time I didn't expect to implement the certificate management on the load balancer, and that's the only reason I added Python code at all.  Everything else is more easily done in shell script.  I don't expect the code of the load balancer to signifcantly grow in the future, by the way, although code has a tendency to do so even if you don't expect it.

**Comment by mtyaka** [comment]
👍 The updates look good. I did not run the integration tests myself, but since Jill already did that and the CircleCI build is green, I think this is good to go.

**Pull request merged**
